### ISSUE TITLE
Ensure AWS environment is healthy upon startup

### ISSUE THREAD

**Issue by katherine-black** [cloudigrade/cloudigrade#738]
In GitLab by @infinitewarp on Sep 28, 2020, 15:31

## Important Note:

We are holding this issue from completing its implementation until we have either a) strong guidance for a solution regarding shared secrets/configs from the platform or b) March 26th has arrived which means we've given several weeks opportunity for conversation and feedback. If we have no new or strong guidance by March 26, let's proceed with whatever "ugly" but working solution we can identify (such as shoving configs in an s3 bucket).

## Summary

As a cloudigrade sysop, I want an automated process to run upon startup (alongside DB migrations) that ensures my AWS environment is appropriately configured and available so that cloudigrade can more seamlessly handle AWS environment changes. Currently, we have manual steps for setting up things in AWS that include defining objects and policies in ECS, SQS, and S3. We should automate their setup and verification upon cloudigrade deployment.

Aside from generally making our lives easier, this will become necessary as we move towards ephemeral environments and if/when the platform requests changing our main AWS account. The latter has been proposed at least once to me as a contingency for occasional but unexpected AWS service outages.

## Acceptance Criteria

- [x] Verify that cloudigrade automatically configures an "empty" AWS account with:
  - [ ] ECS cluster (including Auto Scaling Group and Launch Configurations) for houndigrade
  - [ ] Security Group for the Launch Configuration
  - [ ] SQS queue and S3 bucket with appropriate policies for CloudTrail
  - [ ] ...and anything else relevant we think of during development.
- [x] Verify that cloudigrade automatically REconfigures a "not-empty" AWS account with the same as above
- [x] Verify that cloudigrade alerts Slack after the deployment process finishes, with a brief status of the run
- [ ] QE to rerun IQE regression tests.

## Assumptions and Questions

- Probably no specific QE efforts on this; no automation tests.
- Should we "start small" here by simply migrating and modernizing our ooooooold Ansible scripts that are used for our review environments?
  - Yes, try this!
- If we see that an account appears to be set up but is set up **incorrectly**, should we attempt to repair? Or should we stop and alert as a failure/error?
  - No, we blindly rerun Ansible scripts. See next point.
- What is the order of operations here?
  - 1. We should run Ansible scripts in dry-run/verify mode FIRST.
  - 2. Always alert Sentry (consider adding a new Slack channel) with the results. Future iteration might make this conditional.
  - 3. Regardless of the dry-run/verify results, always run the Ansible scripts to attempt setup.
  - 4. Always alert Sentry (consider adding a new Slack channel) with the results. Future iteration might make this conditional.
  - Yes, this means every deployment likely results in two Sentry alerts (for now).
- Should we also have a bigger discussion about our deployment strategy? Should we move the mid-hook things from the API image to something else?
  - We should (for now) just always run migrations in the mid-hook.
  - Have a discussion with the team if we find any interesting options that are not super quick and easy. Consider opening a new issue for that work if it's bigger.
- Assume the RDS _exists_ and we don't need to set it up.
  - Maybe add an call to check that the DB exists at the configured hostname.
  - What about ephemeral environments? Will they configure and set up RDS for us?
    - For now, let's assume that they will handle this. If we learn that they do not, we will create a future issue to do this ourselves conditionally.
- Should we have some sanity checks during setup that warn us of emptiness (no schema/tables)?
  - Let's not do this now, but consider adding this in the future.
- This means some environment variables should probably go away or be derived from other variables. Perhaps we should always use a standard name plus some kind of configurable environment-specific prefix? We should change:
  - HOUNDIGRADE_AWS_AUTOSCALING_GROUP_NAME
  - HOUNDIGRADE_ECS_CLUSTER_NAME
  - HOUNDIGRADE_RESULTS_QUEUE_NAME
  - HOUNDIGRADE_ECS_FAMILY_NAME
  - S3_BUCKET_NAME
  - CLOUDTRAIL_NAME_PREFIX

**Comment by infinitewarp** [comment]
See also: https://github.com/cloudigrade/houndigrade/issues/108

**Comment by infinitewarp** [comment]
After numerous experiments and discoveries and workarounds to make this process automated in a reasonable fashion, @katherine-black has hit an issue that is a blocker for proceeding with our current design plan.

We want to attach a persistent volume claim to all of our deployed pods and put a config file in that PVC during the deployment mid hook with rotated credentials and various other updated IDs. However, OSD does not appear to have a PVC storage class that actually supports mounting and reading from multiple pods at the same time. We need a PVC that supports Shared Access (RWX), but only Single User (RWO) actually appears to work in OpenShift Dedicated. Attempting to use a PVC with gp2 (the only storage class available) and RWX results in only _one_ pod connecting and all subsequent ones failing.

We just had a conversation with Chris Duryee to demonstrate this problem, and he is reaching out to contacts to see if anyone can help us here.

**Comment by infinitewarp** [comment]
[Response] from Chris:
> for the storage question about gp2, it sounds like the two best options are a.) s3 or b.) stand up a small redis instance. neither of those are great so if you want to brainstorm more options let me know
>
> like maybe you could have a single pod that has only one job which is serving secrets or something
> or rework things to use something that already exists
>
> i know s3 pins you on aws but IMO if we moved off aws there would be a lot of other things we'd have to cut over as well. i dont think this would add a large amount of tech debt

[Response]: https://ansible.slack.com/archives/C8VGAPJNN/p1614632263046200

**Comment by katherine-black** [comment]
Unblocking after a convo w/@infinitewarp

**Comment by infinitewarp** [comment]
To expand on that here, I think we're going to try to resume work on this by cutting our scope down to setting up AWS entities that _don't_ require us remembering any new values since we don't have a reliable shared storage mechanism. So, we likely will not be rotating limited-access keys with this effort, but we should follow up and try to solve that problem in a future issue.

**Comment by infinitewarp** [comment]
After much back-and-forth experimenting and discussion about potential revised definitions and solutions, since we now have a working solution in all of our supported environments, I think I'm happy to call this one done. Thanks for all the hard work, @katherine-black ! 🙂
